### ISSUE TITLE
ModelPruning and PyTorch Profiler incompatibility

### ISSUE THREAD

**Issue by austinmw** [PyTorchLightning/pytorch-lightning#12393]
## 🐛 Bug

When I attempt to use the pruning callback (`ModelPruning('l1_unstructured', amount=0.5)`) in conjunction with `profiler='pytorch'` I get the following error:

> zz2hyoqiv6-algo-1-vqegg  | Traceback (most recent call last):
> zz2hyoqiv6-algo-1-vqegg  |   File "train.py", line 51, in <module>
> zz2hyoqiv6-algo-1-vqegg  | main(args)
> zz2hyoqiv6-algo-1-vqegg  |   File "train.py", line 44, in main
> zz2hyoqiv6-algo-1-vqegg  | else:             trainer.fit(model, dm)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 771, in fit
> zz2hyoqiv6-algo-1-vqegg  | self._call_and_handle_interrupt(
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 722, in _call_and_handle_interrupt
> zz2hyoqiv6-algo-1-vqegg  | return self.strategy.launcher.launch(trainer_fn, *args, trainer=self, **kwargs)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/strategies/launchers/subprocess_script.py", line 93, in launch
> zz2hyoqiv6-algo-1-vqegg  | return function(*args, **kwargs)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 812, in _fit_impl
> zz2hyoqiv6-algo-1-vqegg  | results = self._run(model, ckpt_path=self.ckpt_path)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1174, in _run
> zz2hyoqiv6-algo-1-vqegg  | self._call_setup_hook()  # allow user to setup lightning_module in accelerator environment
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1493, in _call_setup_hook
> zz2hyoqiv6-algo-1-vqegg  | self._call_callback_hooks("setup", stage=fn)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1636, in _call_callback_hooks
> zz2hyoqiv6-algo-1-vqegg  | fn(self, self.lightning_module, *args, **kwargs)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/callbacks/pruning.py", line 378, in setup
> zz2hyoqiv6-algo-1-vqegg  | self._original_layers.setdefault(id_, _LayerRef(data=deepcopy(module), names=[]))
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 172, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  | y = _reconstruct(x, memo, *rv)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 270, in _reconstruct
> zz2hyoqiv6-algo-1-vqegg  | state = deepcopy(state, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 146, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  |     y = copier(x, memo)
> zz2hyoqiv6-algo-1-vqegg  | File "/opt/conda/lib/python3.8/copy.py", line 230, in _deepcopy_dict
> zz2hyoqiv6-algo-1-vqegg  | y[deepcopy(key, memo)] = deepcopy(value, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 172, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  |     y = _reconstruct(x, memo, *rv)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 296, in _reconstruct
> zz2hyoqiv6-algo-1-vqegg  | value = deepcopy(value, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 172, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  |     y = _reconstruct(x, memo, *rv)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 264, in _reconstruct
> zz2hyoqiv6-algo-1-vqegg  | y = func(*args)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 263, in <genexpr>
> zz2hyoqiv6-algo-1-vqegg  | args = (deepcopy(arg, memo) for arg in args)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 146, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  |     y = copier(x, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 237, in _deepcopy_method
> zz2hyoqiv6-algo-1-vqegg  | return type(x)(x.__func__, deepcopy(x.__self__, memo))
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 172, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  |     y = _reconstruct(x, memo, *rv)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 270, in _reconstruct
> zz2hyoqiv6-algo-1-vqegg  |     state = deepcopy(state, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 146, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  | y = copier(x, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 230, in _deepcopy_dict
> zz2hyoqiv6-algo-1-vqegg  |     y[deepcopy(key, memo)] = deepcopy(value, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 172, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  | y = _reconstruct(x, memo, *rv)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 270, in _reconstruct
> zz2hyoqiv6-algo-1-vqegg  | state = deepcopy(state, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 146, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  |     y = copier(x, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 230, in _deepcopy_dict
> zz2hyoqiv6-algo-1-vqegg  |     y[deepcopy(key, memo)] = deepcopy(value, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 172, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  | y = _reconstruct(x, memo, *rv)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 296, in _reconstruct
> zz2hyoqiv6-algo-1-vqegg  |     value = deepcopy(value, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 172, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  | y = _reconstruct(x, memo, *rv)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 270, in _reconstruct
> zz2hyoqiv6-algo-1-vqegg  | state = deepcopy(state, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 146, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  |     y = copier(x, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 210, in _deepcopy_tuple
> zz2hyoqiv6-algo-1-vqegg  | y = [deepcopy(a, memo) for a in x]
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 210, in <listcomp>
> zz2hyoqiv6-algo-1-vqegg  |     y = [deepcopy(a, memo) for a in x]
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 146, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  | y = copier(x, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 210, in _deepcopy_tuple
> zz2hyoqiv6-algo-1-vqegg  |     y = [deepcopy(a, memo) for a in x]
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 210, in <listcomp>
> zz2hyoqiv6-algo-1-vqegg  | y = [deepcopy(a, memo) for a in x]
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 172, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  |     y = _reconstruct(x, memo, *rv)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 270, in _reconstruct
> zz2hyoqiv6-algo-1-vqegg  | state = deepcopy(state, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 146, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  |     y = copier(x, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 230, in _deepcopy_dict
> zz2hyoqiv6-algo-1-vqegg  | y[deepcopy(key, memo)] = deepcopy(value, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 172, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  |     y = _reconstruct(x, memo, *rv)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 270, in _reconstruct
> zz2hyoqiv6-algo-1-vqegg  | state = deepcopy(state, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 146, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  |     y = copier(x, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 230, in _deepcopy_dict
> zz2hyoqiv6-algo-1-vqegg  | y[deepcopy(key, memo)] = deepcopy(value, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 146, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  | y = copier(x, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 205, in _deepcopy_list
> zz2hyoqiv6-algo-1-vqegg  |     append(deepcopy(a, memo))
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 172, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  |     y = _reconstruct(x, memo, *rv)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 270, in _reconstruct
> zz2hyoqiv6-algo-1-vqegg  | state = deepcopy(state, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 146, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  |     y = copier(x, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 230, in _deepcopy_dict
> zz2hyoqiv6-algo-1-vqegg  |     y[deepcopy(key, memo)] = deepcopy(value, memo)
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/copy.py", line 161, in deepcopy
> zz2hyoqiv6-algo-1-vqegg  | rv = reductor(4)
> zz2hyoqiv6-algo-1-vqegg  | TypeError: cannot pickle '_io.TextIOWrapper' object
> zz2hyoqiv6-algo-1-vqegg  | Exception ignored in: <function BaseProfiler.__del__ at 0x7fd53b5c7940>
> zz2hyoqiv6-algo-1-vqegg  | Traceback (most recent call last):
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/profiler/base.py", line 199, in __del__
> zz2hyoqiv6-algo-1-vqegg  | File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/profiler/pytorch.py", line 509, in teardown
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/profiler/pytorch.py", line 494, in _delete_profilers
> zz2hyoqiv6-algo-1-vqegg  |   File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/profiler/pytorch.py", line 489, in _cache_functions_events
> zz2hyoqiv6-algo-1-vqegg  | File "/opt/conda/lib/python3.8/site-packages/torch/profiler/profiler.py", line 382, in events
> zz2hyoqiv6-algo-1-vqegg  | AssertionError:
> zz2hyoqiv6-algo-1-vqegg  | 2022-03-21 12:47:50,301 sagemaker-training-toolkit ERROR    Reporting training FAILURE
> zz2hyoqiv6-algo-1-vqegg  | 2022-03-21 12:47:50,301 sagemaker-training-toolkit ERROR    ExecuteUserScriptError:
> zz2hyoqiv6-algo-1-vqegg  | ExitCode 1
> zz2hyoqiv6-algo-1-vqegg  | ErrorMessage "TypeError: cannot pickle '_io.TextIOWrapper' object
> zz2hyoqiv6-algo-1-vqegg  |  Exception ignored in: <function BaseProfiler.__del__ at 0x7fd53b5c7940> Traceback (most recent call last):   File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/profiler/base.py", line 199, in __del__ File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/profiler/pytorch.py", line 509, in teardown   File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/profiler/pytorch.py", line 494, in _delete_profilers   File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/profiler/pytorch.py", line 489, in _cache_functions_events File "/opt/conda/lib/python3.8/site-packages/torch/profiler/profiler.py", line 382, in events AssertionError:"

When I comment out `profiler='pytorch'` the fit call runs fine.

### To Reproduce
I have not yet been able to successfully reproduce with the BoringModel. If I manage to I will update this issue.

### Expected behavior

No error.

### Environment

You can also fill out the list below manually.
-->

- PyTorch Lightning Version (e.g., 1.5.0): mainline
- PyTorch Version (e.g., 1.10): 1.10
- Python version (e.g., 3.9): 3.8
- OS (e.g., Linux): Ubuntu
- CUDA/cuDNN version: 11.3
- GPU models and configuration: V100
- How you installed PyTorch (`conda`, `pip`, source): pip git mainline

**Comment by carmocca** [comment]
Is this still an issue in the latest release? Any progress on the reproducible script?

**Comment by austinmw** [comment]
Hey, apologies, but haven't reproduced this lately.

**Comment by deependujha** [comment]
Hi! I spent some time investigating this issue. Here's what I found.

> **TL;DR:** I can no longer reproduce this issue on current versions (PyTorch Lightning 2.6.5 / PyTorch 2.9.1 / Python 3.12). Closing this as resolved.

## Reproduction

I tested both the originally reported `ModelPruning + profiler="pytorch"` combination, as well as the `resample_parameters=True` code path (which performs an additional `deepcopy` internally):

```python
import copy
import traceback

import torch
import torch.nn as nn
import torch.nn.utils.prune as prune
import pytorch_lightning as pl
from pytorch_lightning.callbacks import ModelPruning


class SimpleModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.layer1 = nn.Linear(32, 64)
        self.layer2 = nn.Linear(64, 10)

    def forward(self, x):
        return self.layer2(torch.relu(self.layer1(x)))

    def training_step(self, batch, batch_idx):
        x, y = batch
        return nn.functional.cross_entropy(self(x), y)

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=1e-3)


class RandomDataModule(pl.LightningDataModule):
    def setup(self, stage=None):
        self.d = torch.utils.data.TensorDataset(
            torch.randn(64, 32), torch.randint(0, 10, (64,))
        )

    def train_dataloader(self):
        return torch.utils.data.DataLoader(self.d, batch_size=32)


print("Test 1: ModelPruning + profiler='pytorch'")
cb = ModelPruning("l1_unstructured", amount=0.5, make_pruning_permanent=False)
trainer = pl.Trainer(
    max_epochs=2,
    callbacks=[cb],
    profiler="pytorch",
    accelerator="auto",
    enable_checkpointing=False,
    enable_progress_bar=True,
)

try:
    trainer.fit(SimpleModel(), RandomDataModule())
    print("PASS: training completed successfully")
except Exception as e:
    print(f"FAIL: {type(e).__name__}: {e}")
    traceback.print_exc()


print("\nTest 2: ModelPruning + profiler + resample_parameters=True")
cb2 = ModelPruning(
    "l1_unstructured",
    amount=0.5,
    make_pruning_permanent=False,
    resample_parameters=True,
)
trainer2 = pl.Trainer(
    max_epochs=2,
    callbacks=[cb2],
    profiler="pytorch",
    accelerator="auto",
    enable_checkpointing=False,
    enable_progress_bar=True,
)

try:
    trainer2.fit(SimpleModel(), RandomDataModule())
    print("PASS: training completed successfully")
except Exception as e:
    print(f"FAIL: {type(e).__name__}: {e}")
    traceback.print_exc()
```

Both tests completed successfully on:

* PyTorch Lightning 2.6.5
* PyTorch 2.9.1
* Python 3.12

## Root cause analysis

I also looked through the repository history to see whether a specific change fixed this behavior, but couldn't identify a targeted fix.

Based on the original stack trace, the failure happened during `ModelPruning.setup()`, where Lightning performs a `deepcopy(module)`:

```text
File ".../callbacks/pruning.py", line 378, in setup
    self.original_layers.setdefault(id, _LayerRef(data=deepcopy(module), names=[]))

TypeError: cannot pickle '_io.TextIOWrapper' object
```

The `TypeError: cannot pickle '_io.TextIOWrapper' object` is raised when Python attempts to pickle (or `deepcopy`) an object graph containing an open file handle.

My best hypothesis is that, at the time, enabling `profiler="pytorch"` caused the profiler to create its output file before `ModelPruning.setup()` executed. Since the profiler also registered forward hooks on the model, `deepcopy(module)` likely traversed those hook references back through the `Trainer` to the profiler, eventually reaching the profiler's open output file (`_output_file`), which isn't picklable.

https://github.com/Lightning-AI/pytorch-lightning/blob/fe6b1cc4e80ae0396e2e404c16e6b6968ad5437e/src/lightning/pytorch/trainer/trainer.py#L1171-L1176

https://github.com/Lightning-AI/pytorch-lightning/blob/fe6b1cc4e80ae0396e2e404c16e6b6968ad5437e/src/lightning/pytorch/profilers/pytorch.py#L83-L95

Interestingly, the current profiler implementation still contains the following comment during teardown:

```python
self._output_file = None  # can't pickle TextIOWrapper
```

https://github.com/Lightning-AI/pytorch-lightning/blob/fe6b1cc4e80ae0396e2e404c16e6b6968ad5437e/src/lightning/pytorch/profilers/profiler.py#L146-L155

which suggests this interaction was known.

On current versions, however, this no longer happens. The `deepcopy` performed by `ModelPruning.setup()` completes successfully before profiler hooks are added to the copied object graph, so the unpicklable file handle is never reached. (It happens in the `start` method of `PyTorchProfiler` now, which is called after `setup (deepcopy of ModelPruning)`.

---

Since this issue has been inactive for more than four years and I can no longer reproduce it on supported versions, I'm going to close it.

If anyone encounters the same problem on a current release, please feel free to open a new issue (or reopen this one) with a minimal reproducible example.
